x/feedist/handler.go

```diff
@@ -0,0 +1,141 @@
+	return res, nil
+}
+
+func (h *distributeHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	rev, err := h.bucket.GetRevenue(db, msg.RevenueID)
+	if err != nil {
+		return res, err
+	}
+
+	if err := distribute(ctx, db, rev); err != nil {
+		return res, errors.Wrap(err, "cannot distribute")
+	}
+	return res, nil
+}
+
+func (h *distributeHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*DistributeMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get message")
+	}
+	msg, ok := rmsg.(*DistributeMsg)
+	if !ok {
+		return nil, errors.InvalidMsgErr.New("unknown transaction type")
+	}
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+	return msg, nil
+}
+
+type updateRevenueHandler struct {
+	auth   x.Authenticator
+	bucket *RevenueBucket
+}
+
+func (h *updateRevenueHandler) Check(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.CheckResult, error) {
+	var res weave.CheckResult
+	if _, err := h.validate(ctx, db, tx); err != nil {
+		return res, err
+	}
+	res.GasAllocated += updateRevenueCost
+	return res, nil
+}
+
+func (h *updateRevenueHandler) Deliver(ctx weave.Context, db weave.KVStore, tx weave.Tx) (weave.DeliverResult, error) {
+	var res weave.DeliverResult
+	msg, err := h.validate(ctx, db, tx)
+	if err != nil {
+		return res, err
+	}
+
+	rev, err := h.bucket.GetRevenue(db, msg.RevenueID)
+	if err != nil {
+		return res, err
+	}
+
+	// Before updating the revenue all funds must be distributed. Only a
+	// revenue with no funds can be updated, so that recipients trust us.
+	// Otherwise an admin could change who receives the money without the
+	// previously selected recepients ever being paid.
+	if err := distribute(ctx, db, rev); err != nil {
+		return res, errors.Wrap(err, "cannot distribute")
+	}
+
+	rev.Recipients = msg.Recipients
+	obj := orm.NewSimpleObj(msg.RevenueID, rev)
+	if err := h.bucket.Save(db, obj); err != nil {
+		return res, errors.Wrap(err, "cannot save")
+	}
+	return res, nil
+}
+
+func (h *updateRevenueHandler) validate(ctx weave.Context, db weave.KVStore, tx weave.Tx) (*UpdateRevenueMsg, error) {
+	rmsg, err := tx.GetMsg()
+	if err != nil {
+		return nil, errors.Wrap(err, "cannot get message")
+	}
+	msg, ok := rmsg.(*UpdateRevenueMsg)
+	if !ok {
+		return nil, errors.InvalidMsgErr.New("unknown transaction type")
+	}
+	if err := msg.Validate(); err != nil {
+		return msg, err
+	}
+	return msg, nil
+}
+
+// distribute split the funds stored under the revenue address and distribute
+// them according to recipients proportions. When successful, revenue account
+// has no funds left after this call.
+func distribute(ctx weave.Context, db weave.KVStore, rev *Revenue) error {
+	var chunks int32
+	for _, r := range rev.Recipients {
+		chunks += r.Weight
+	}
+
+	panic("todo")
```

Comment:
@ethanfrey how can I get the amount stored under the revenue address? I thought that money moving operations could be done with the [coins controller](https://github.com/iov-one/weave/blob/master/x/cash/controller.go#L12-L17), but it only can move coins - it does not allow to check the state of one.

Reply:
That is a good question, and if not available, we may have to expose that in the CoinsController. That or pass in the Bucket itself, but I'd rather not.

Reply:
Yes, there is no way other than those two...
Please add it to `BaseController` and `Controller` in cash.
This may mean adding the same functionality to namecoin, so the method can be generic, like

`GetBalance(addr weave.Address) x.Coins`

I've been thinking about such a while ago, but I guess I never did it